### PR TITLE
feat: Add WITH KEY to show tag keys

### LINKED ISSUE #15993: Add WITH syntax to SHOW TAG KEYS
The SHOW TAG KEYS meta query does not appear to have the WITH syntax that exists with two of the other meta querys: SHOW MEASUREMENTS and SHOW TAG VALUES.

Example: 
SHOW MEASUREMENTS [ON <database_name>] __[WITH MEASUREMENT <regular_expression>]__ [WHERE <tag_key> <operator> ['<tag_value>' | <regular_expression>]] [LIMIT_clause] [OFFSET_clause]

SHOW TAG VALUES [ON <database_name>][FROM_clause] __WITH KEY [ [<operator> "<tag_key>" | <regular_expression>]__ | [IN ("<tag_key1>","<tag_key2")]] [WHERE <tag_key> <operator> ['<tag_value>' | <regular_expression>]] [LIMIT_clause] [OFFSET_clause]

__Current behavior:__
If I want to explore/examine TAG KEYS using regular expressions, I must dump all the tag keys out to a file and then use *nix tooling to extract the elements I'm looking for.  Why does this happen?  If I have a database with a large number of measurements and I'm supporting a large number of users who are writing in data, I don't always know where to look for "rouge tag keys".  Being able to run regular expressions using the WITH syntax would be valuable so that I can use the database itself to complete this exploration.

__Desired behavior:__
Add similar syntax from the SHOW MEASUREMENTS and SHOW TAG VALUES.

Proposed:
SHOW TAG KEYS [ON <database_name>] [FROM_clause] __[WITH KEY <regular_expression>]__ [WHERE <tag_key> <operator> ['<tag_value>' | <regular_expression>]] [LIMIT_clause] [OFFSET_clause]

### PR DESCRIPTION
**Pull request by lesam**
Closes #15993

- [ ] [CHANGELOG.md](https://github.com/influxdata/influxdb/blob/master/CHANGELOG.md) updated with a link to the PR (not the Issue)
- [ ] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [ ] Rebased/mergeable
- [ ] Tests pass
- [ ] http/swagger.yml updated (if modified Go structs or API)
- [ ] Feature flagged (if modified API)
- [ ] Documentation updated or issue created (provide link to issue/pr)
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: tests/server_test.go [hunk 1]
```diff
@@ -7983,6 +7983,24 @@ func TestServer_Query_ShowTagKeys(t *testing.T) {
 			exp:     `{"results":[{"statement_id":0,"series":[{"name":"cpu","columns":["tagKey"],"values":[["host"],["region"]]},{"name":"gpu","columns":["tagKey"],"values":[["host"],["region"]]}]}]}`,
 			params:  url.Values{"db": []string{"db0"}},
 		},
+		// TODO: WITH KEY + rewriting instead of _tagKey =
```
Comment (lesam) [inline tests/server_test.go:7986]: Note the TODO

File: tsdb/shard.go [hunk 2]
```diff
@@ -801,26 +801,6 @@ func (s *Shard) MeasurementNamesByRegex(re *regexp.Regexp) ([][]byte, error) {
...
-// provided expression.
-func (s *Shard) MeasurementTagKeyValuesByExpr(auth query.Authorizer, name []byte, key []string, expr influxql.Expr, keysSorted bool) ([][]string, error) {
-	index, err := s.Index()
-	if err != nil {
-		return nil, err
-	}
-	indexSet := IndexSet{Indexes: []Index{index}, SeriesFile: s.sfile}
-	return indexSet.MeasurementTagKeyValuesByExpr(auth, name, key, expr, keysSorted)
-}
-
```
Comment (dgnorton) [inline tsdb/shard.go:823]: Is this dead code?

Comment (lesam) [reply tsdb/shard.go:823]: Yes

File: tsdb/index.go [hunk 3]
```diff
@@ -2443,43 +2443,40 @@ func (is IndexSet) matchTagValueNotEqualNotEmptySeriesIDIterator(name, key []byt
...
-// TagValuesByKeyAndExpr retrieves tag values for the provided tag keys.
+// tagValuesByKeyAndExpr retrieves tag values for the provided tag keys.
 //
-// TagValuesByKeyAndExpr returns sets of values for each key, indexable by the
+// tagValuesByKeyAndExpr returns sets of values for each key, indexable by the
 // position of the tag key in the keys argument.
 //
 // N.B tagValuesByKeyAndExpr relies on keys being sorted in ascending
 // lexicographic order.
-func (is IndexSet) TagValuesByKeyAndExpr(auth query.Authorizer, name []byte, keys []string, expr influxql.Expr, fieldset *MeasurementFieldSet) ([]map[string]struct{}, error) {
```
Comment (lesam) [inline tsdb/index.go:2453]: Dead

File: tsdb/store.go [hunk 4]
```diff
@@ -1663,7 +1656,7 @@ func (s *Store) TagKeys(auth query.Authorizer, shardIDs []uint64, cond influxql.
 	for _, name := range names {
 
 		// Build keyset over all indexes for measurement.
-		tagKeySet, err := is.MeasurementTagKeysByExpr(name, nil)
+		tagKeySet, err := is.MeasurementTagKeysByExpr(name, tagKeyExpr)
```
Comment (lesam) [inline tsdb/store.go:1659]: Same behaviour as 'show tag value with key'

File: tsdb/store.go [hunk 5]
```diff
@@ -1845,7 +1859,7 @@ func (s *Store) TagValues(auth query.Authorizer, shardIDs []uint64, cond influxq
 	// filter.
 	for _, name := range names {
 		// Determine a list of keys from condition.
-		keySet, err := is.MeasurementTagKeysByExpr(name, cond)
+		keySet, err := is.MeasurementTagKeysByExpr(name, tagKeyExpr)
```
Comment (lesam) [inline tsdb/store.go:1862]: This worked before because we *always* had a `_tagKey` field (enforced by the parser). But if you don't have any binary expressions with `_tagKey` but you do have a non-nil expression this returns an empty set.

File: tsdb/store.go [hunk 6]
```diff
@@ -1900,136 +1914,24 @@ func (s *Store) TagValues(auth query.Authorizer, shardIDs []uint64, cond influxq
...
-		// An invariant is that there can't be more than n instances of tag
-		// key value pairs for a given measurement, where n is the number of
-		// shards.
-		if got, exp := j-i+1, len(is.Indexes); got > exp {
-			return nil, fmt.Errorf("unexpected results returned engine. Got %d measurement sets for %d shards", got, exp)
-		}
-
-		nextResult := mergeTagValues(idxBuf, allResults[i:j+1]...)
-		i = j + 1
+	for _, r := range allResults {
```
Comment (lesam) [inline tsdb/store.go:1917]: Probably this used to be needed before iterators over IndexSet? Now the merge is done within the iterator.

File: tsdb/index.go [hunk 7]
```diff
@@ -2512,6 +2509,18 @@ func (is IndexSet) tagValuesByKeyAndExpr(auth query.Authorizer, name []byte, key
 			break
 		}
 
+		if e.Expr != nil {
+			// We don't yet have code that correctly processes expressions that
+			// seriesByExprIterator doesn't handle
+			lit, ok := e.Expr.(*influxql.BooleanLiteral)
+			if !ok {
+				return nil, fmt.Errorf("Expression too complex for metaquery: %v", e.Expr)
```
Comment (lesam) [inline tsdb/index.go:2517]: Note this new error condition.

File: tsdb/exprs.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+			passSlice = append(passSlice, e)
+		} else {
+			failSlice = append(failSlice, e)
+		}
+	}
+
+	passExpr = ExprsToConjunction(passSlice...)
+	failExpr = ExprsToConjunction(failSlice...)
+	return passExpr, failExpr, nil
+}
```
Comment (benbjohnson) [inline tsdb/exprs.go:90]: The function looks fine but we use the term "partition" in a lot of places in the code and this partitioning seems unrelated to any of those. I'm not sure what better word would work but it's something to consider.

**Review by benbjohnson — DISMISSED** [review]

File: tsdb/exprs.go (hunk 8 above)
Comment (lesam) [reply tsdb/exprs.go:90]: Maybe. We use the same term in the flux code though (PartitionPredicate)

File: tsdb/store.go [hunk 9]
```diff
@@ -1935,150 +1949,35 @@ func (s *Store) TagValues(ctx context.Context, auth query.Authorizer, shardIDs [
...
-	// The next stage is to merge the tagValue results for each shard's measurements.
-	var i, j int
-	// Used as a temporary buffer in mergeTagValues. There can be at most len(shards)
-	// instances of tagValues for a given measurement.
-	idxBuf := make([][2]int, 0, len(is.Indexes))
-	for i < len(allResults) {
+	result := make([]TagValues, 0, maxMeasurements)
+	for _, r := range allResults {
 		// check for timeouts
 		select {
```
Comment (lesam) [inline tsdb/store.go:1958]: Fixed this merge conflict with @davidby-influx 's PR merged yesterday.

File: query/statement_rewriter.go [hunk 10]
```diff
@@ -372,9 +381,24 @@ func rewriteShowTagValuesCardinalityStatement(stmt *influxql.ShowTagValuesCardin
...
+			condition = &influxql.BinaryExpr{
+				LHS: &influxql.ParenExpr{Expr: condition},
+				RHS: &influxql.ParenExpr{Expr: tagExpr},
+				Op:  influxql.AND,
+			}
+		} else {
+			// condition is nil, replace with tagExpr
+			condition = tagExpr
+		}
+		// other cases: tagExpr is nil, so condition is already set correctly
```
Comment (dgnorton) [inline query/statement_rewriter.go:397]: Should this comment go below the closing `}` for `if tagExpr != nil {`?

File: tsdb/exprs.go [hunk 11]
```diff
@@ -0,0 +1,90 @@
+package tsdb
```
Comment (dgnorton) [inline tsdb/exprs.go:1]: Weren't some of the functions in this file merged into the influxql repo? Do we still need this new file in influxdb?

File: tsdb/index.go (hunk 7 above)
Comment (dgnorton) [inline tsdb/index.go:2517]: Wondering if there is a more informative error message we could give the user here. "Meta queries don't support..."?

Comment (lesam) [reply tsdb/index.go:2517]: I'm not sure what the exact rules are unfortunately, and I'm worried this would be hard to maintain if we add more support in the various places we partially evaluate the query.

File: query/statement_rewriter.go (hunk 10 above)
Comment (lesam) [reply query/statement_rewriter.go]: Moved it above the if block.

**Review by dgnorton — APPROVED** [review]

**Review by luismedicina23 — COMMENTED** [review]
@lesam 
Im sorry I'm new

File: go.mod [hunk 12]
```diff
@@ -16,7 +16,7 @@ require (
 	github.com/golang/snappy v0.0.1
 	github.com/google/go-cmp v0.5.0
 	github.com/influxdata/flux v0.65.0
-	github.com/influxdata/influxql v1.1.1-0.20200828144457-65d3ef77d385
```
Comment (luismedicina23) [inline go.mod:19]: 19- tsdb/store_internal_test.go

**Review by luismedicina23 — COMMENTED** [review]
@lesam 

Im trying

**Comment by luismedicina23** [comment]
@Main

Im new and licensing 

Im Brazilian guy

Sry

**Review by luismedicina23 — COMMENTED** [review]
@MIn 
I'm begin

**Review by luismedicina23 — COMMENTED** [review]
@luismedicina23

**Review by luismedicina23 — COMMENTED** [review]
@luismedicina23

File: tests/server_test.go [hunk 13]
```diff
@@ -7983,6 +7983,23 @@ func TestServer_Query_ShowTagKeys(t *testing.T) {
...
+		},
+		&Query{
+			name:    "show tag keys from with key",
+			command: "SHOW TAG KEYS FROM cpu WITH KEY = host",
+			exp:     `{"results":[{"statement_id":0,"series":[{"name":"cpu","columns":["tagKey"],"values":[["host"]]}]}]}`,
+			params:  url.Values{"db": []string{"db0"}},
+		},
+		&Query{
+			name:    "show tag keys from regex with key in",
+			command: "SHOW TAG KEYS FROM /[cg]pu/ WITH KEY IN (host, region) ",
```
Comment (lesam) [inline tests/server_test.go:7999]: good example query
